src/decode.rs

```diff
@@ -408,6 +405,43 @@ pub fn decode_value<I: Iterator<Item = bool>>(
     Ok(value)
 }
 
+/// Decode a value from bits, of the form 2^exp
+///
+/// # Panics
+///
+/// Panics if `exp` itself is not a power of 2
+pub fn decode_power_of_2<I: Iterator<Item = bool>>(
+    iter: &mut I,
+    exp: usize,
+) -> Result<Value, Error> {
+    assert_eq!(exp.count_ones(), 1, "exp must be a power of 2");
+
+    struct StackElem {
+        value: Value,
+        width: usize,
+    }
+
+    let mut stack = Vec::with_capacity(32);
+    for _ in 0..exp {
+        // Read next bit
+        let bit = Value::u1(u8::from(iter.next().ok_or(Error::EndOfStream)?));
+        stack.push(StackElem {
+            value: bit,
+            width: 1,
+        });
+
+        while stack.len() >= 2 && stack[stack.len() - 1].width == stack[stack.len() - 2].width {
+            let right = stack.pop().unwrap();
+            let left = stack.pop().unwrap();
+            stack.push(StackElem {
+                value: Value::Prod(Box::new(left.value), Box::new(right.value)),
+                width: left.width * 2,
+            });
+        }
+    }
+    Ok(stack.pop().unwrap().value)
+}
```

Comment:
795cd3383cda71bed4aae43bbf4101d7540bbd6c: Why should we use this specialized function? We have general functions and caching in the context that should be faster. Let's keep the total number of functions low.

Reply:
Because this is doesn't need a context object or precomputed values. This completely removes the `context` parameter from `decode_program`.

Adding a whole extra parameter to all decoding logic to save a single function that users don't need to care about, does not seem like a reasonable tradeoff to me.